File: general-solving/hackerrank/regex/07_re-applications_05_ip-address-validation.cpp
/***************************************************************************
* Title : IP Address Validation
* URL   : https://www.hackerrank.com/challenges/ip-address-validation/
* Date  : 2018-04-06
* Author: Atiq Rahman
* Comp  : O(n) to do multiple iterations, space O(n) to store tokens
* Status: Accepted
* Notes : This is solution for bloomberg version which only validates IPv4
*   For hackerrank version which includes validation for IPv6 check the C#
*   solution
*   If input string is valid IPv4 print IPv4
*   Otherwise print Invalid
*   Uses C++ istringstream to tokenize;
*   Example ref, https://atiqcs.wordpress.com/2018/01/27/cpp-fundamentals/
*   Example input at the bottom of the source
*   The interviewer suggested 'cerr' for debugging errors.
* meta  : tag-easy, tag-company-bloomberg, tag-string, tag-tokenize
***************************************************************************/
#include <sstream>
#include <string>
#include <vector>
#include <fstream>
#include <iostream>
#include <algorithm>    // find_if
#include <cctype>       // isdigit

class IPValidation {
private:
  // ref: How to determine if a string is a number with C++ 11?
  // https://stackoverflow.com/q/4654636
  bool is_number(const std::string& s)
  {
    return !s.empty() && std::find_if(s.begin(),
      s.end(), [](char c) { return !std::isdigit(c); }) == s.end();
  }

  // method 1: using std::stoi
  // ref: http://en.cppreference.com/w/cpp/string/basic_string/stol
  int convert_string_to_number(const std::string& s)
  {
    int val = -1;
    try {
      val = std::stoi(s);
    }
    catch (const std::invalid_argument& e) {
    }
    catch (const std::out_of_range& e) {
    }
    // when val = -1 caller function gets the meaning
    return val;
  }
  // method 2: using std::istringstream; need to test
  /* int convert_string_to_number(const std::string& s)
  {
    std::istringstream ss(s);
    int val;
    if (ss >> val)
    return val;
    return -1;
  }*/

  bool isValid(std::string ip) {
    // tokenize
    std::istringstream ss(ip);
    std::string token;
    int count = 0;
    char delim = '.';

    while (std::getline(ss, token, delim)) {
      count++;
      // early termination
      if (count > 4)
        return false;
      // validate if token is a number
      //if (is_number(token) == false)
      //return false;
      // valid input std::string, verify for ip token
      int val = convert_string_to_number(token);
      if (val < 0 || val > 255)
        return false;
    }
    return count == 4;
  }

  std::vector<std::string> checkIPs(std::vector<std::string> ip_array) {
    std::vector <std::string> result;
    for (auto ip : ip_array)
      result.push_back(isValid(ip) ? "IPv4" : "Invalid");
    return result;
  }

public:
  void Run(std::string inputFile) {
    // Handle IO
    std::ofstream fout(inputFile);
    std::vector <std::string> res;

    int _ip_array_size = 0;
    std::cin >> _ip_array_size;
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    std::vector<std::string> _ip_array;
    std::string _ip_array_item;
    for (int _ip_array_i = 0; _ip_array_i<_ip_array_size; _ip_array_i++) {
      std::getline(std::cin, _ip_array_item);
      _ip_array.push_back(_ip_array_item);
    }

    res = checkIPs(_ip_array);
    for (size_t res_i = 0; res_i < res.size(); res_i++) {
      fout << res[res_i] << std::endl;;
    }

    fout.close();
  }
};

int main() {
  IPValidation demo;
  demo.Run("validation.txt");
  return 0;
}

/*
cerr examples,
cerr << "count " << count << "for input "<< ip << endl;
cerr << "count " << count << "for input "<< ip << endl;

Example input,
3
This is a text.
123.56.98.75
192.168.70.152

Output goes into specified 'validation.txt'
*/

File: general-solving/hackerrank/regex/07_re-applications_05_ip-address-validation.cs
/***************************************************************************
* Title : IP Address Validation
* URL   : https://www.hackerrank.com/challenges/ip-address-validation/
* Date  : 2018-04-07
* Author: Atiq Rahman
* Comp  : O(n) to do multiple iterations, space O(n) to store tokens
* Status: Accepted
* Notes : Solved based on bloomberg version which only validates IPv4
*   This one validates IPv6 as well
*   If input string is neither ipv4 nor ipv6 print Invalid
*   Uses C++ istringstream to tokenize;
*   Example ref, https://atiqcs.wordpress.com/2018/01/27/cpp-fundamentals/
*   Example input at the bottom of the source
*   
*   demonstrates hex number string to int conversion
*   ref, Int32.TryParse Method (String, NumberStyles, IFormatProvider, Int32)
*   https://msdn.microsoft.com/en-us/library/zf50za27.aspx
* meta  : tag-easy, tag-company-bloomberg, tag-string, tag-tokenize
***************************************************************************/
using System;
using System.Globalization;   // hex convert

public class IPValidation {
  private bool IsIPv4(string ip)
  {
    string[] tokens = ip.Split('.');
    if (tokens.Length != 4)
      return false;
    foreach (var token in tokens)
      if (int.TryParse(token, out int n) == false || n>255 || n<0)
        return false;
    return true;
  }

  private bool IsIPv6(string ip)
  {
    string[] tokens = ip.Split(':');
    if (tokens.Length != 8)
      return false;
    int n = -1;
    foreach (var token in tokens)
      if (token.Length>4 || int.TryParse(token, NumberStyles.HexNumber, new
        CultureInfo("en-US"), out n) == false)
        return false;
    return true;
  }

  public void Run()
  {
    int T = int.Parse(Console.ReadLine());
    while (T-- > 0) {
      string str = Console.ReadLine();
      if (IsIPv4(str))
        Console.WriteLine("IPv4");
      else if (IsIPv6(str))
        Console.WriteLine("IPv6");
      else
        Console.WriteLine("Neither");
    }
  }
}

public class HK_Solution {
  public static void Main() {
    IPValidation demo = new IPValidation();
    demo.Run();
  }
}

/*
4
This line has junk text.
121.18.19.20
2001:0db8:0000:0000:0000:ff00:0042:8329
2001:0db8:0000:0000:0000:fff00:0042:8329
*/
